docs/cookbook/invoices/generating-invoice-after-payment.rst

```diff
@@ -0,0 +1,169 @@
+
+The invoicing plugin lets you generate and download PDF invoices regarding orders. In its default behavior the invoice
+is generated right after the customer creates and places the order.
+In this cookbook, we will describe how to change this behavior, so the invoice will be created right after the order would be paid.
+
+Why would you customize the invoice generation?
+-----------------------------------------------
+
+In the default case the order should be immutable, so any other data should not be changed. But let's say that your shop is customized with some logic
+which is not out of the box. Maybe one of these changes will let you change the order after it is placed?
+In this case, it would be better to have an invoice generated after a particular step (in the case of this cookbook - after the order is paid).
+Getting started
+---------------
+
+Before you start, make sure that you have:
+
+#. `Sylius/InvoicingPlugin <https://github.com/Sylius/InvoicingPlugin/>`_ installed.
+#. `Wkhtmltopdf <https://wkhtmltopdf.org/>`_ package installed, because most of pdf generation is done with it.
+
+How to customize the invoice generation?
+----------------------------------------
+
+The concept is quite straightforward and it is based on Symfony events and event listeners.
+We need to create 3 classes and declare them in config files.
+Let's start first with EventListeners that will override the default ones:
+
+**1.** The ``OrderPaymentPaidListener`` will create an invoice and check if it does not exist:
+
+.. code-block:: php
+
+    <?php
+
+    declare(strict_types=1);
+
+    namespace App\EventListener;
+
+    use Sylius\InvoicingPlugin\Command\SendInvoiceEmail;
+    use Sylius\InvoicingPlugin\Creator\InvoiceCreatorInterface;
+    use Sylius\InvoicingPlugin\Event\OrderPaymentPaid;
+    use Sylius\InvoicingPlugin\Exception\InvoiceAlreadyGenerated;
+    use Symfony\Component\Messenger\MessageBusInterface;
+
+    final class OrderPaymentPaidListener
+    {
+        /** @var InvoiceCreatorInterface */
+        private $invoiceCreator;
+
+        /** @var MessageBusInterface */
+        private $commandBus;
+
+        public function __construct(InvoiceCreatorInterface $invoiceCreator, MessageBusInterface $commandBus)
+        {
+            $this->invoiceCreator = $invoiceCreator;
+            $this->commandBus = $commandBus;
+        }
+
+        public function __invoke(OrderPaymentPaid $event): void
+        {
+            try {
+                $this->invoiceCreator->__invoke($event->orderNumber(), $event->date());
+            } catch (InvoiceAlreadyGenerated $exception) {
+            }
+
+            $this->commandBus->dispatch(new SendInvoiceEmail($event->orderNumber()));
+        }
+    }
+
+**2.** The ``NoInvoiceOnOrderPlacedListener`` will not do anything (what a lazy boy), but as we are changing the behavior, it still has very important role:
```

Comment:
you can also remove this service from the container, but I am not sure if it will be a better or a different solution :)

Reply:
Well, that's true but I think this is the easier solution ( and much consistent with other cookbooks).